### ISSUE TITLE
Meetup.com REST to GraphQL API Migration / Conversion - Requires Pro Subscription

### ISSUE THREAD

**Issue by allella** [hackgvl/hackgreenville-com#212]
UPDATE (April 15, 2024) : As [noted below](https://github.com/hackgvl/hackgreenville-com/issues/212#issuecomment-2058154521), RefactorGVL now has a Pro account and we would need to stay within their Terms of Service, including having a privacy policy defined before [applying for an Oauth token through RefactorGVL's admin account](https://help.meetup.com/hc/en-us/articles/360028901812-Using-Meetup-s-API).

The [throttling of requests is on a points system](https://www.meetup.com/api/guide/#p05-rate-limiting).

Also, see https://github.com/hackgvl/hackgreenville-com/issues/172 for examples of using the GraphQL Playground and a couple other points.

----------------------(Notes from Aug 2022)----------------------------------------------------

@ramonaspence 

It's unclear if/when Meetup will force everyone to [migrate from REST to GraphQL](https://www.meetup.com/api/migrating/#graphQl-migrating) and their [API docs are now GraphQL specific](https://www.meetup.com/api/general/).  I figure they'd list a date if they were planning to kill off the REST API, but you never know.  Plus, this would be an opportunity for both of use to use OAuth 2 and GraphQL on a real project. Both seem to be the latest tech for such things.

We'd probably use [JWT Flow](https://www.meetup.com/api/authentication/#p04-jwt-flow-section) to authenticate over OAuth 2. I think each developer could create their own "OAuth client" with a unique AUTHORIZED_MEMBER_ID and CLIENT_ID which we'd likely store in the local .env / environment / config.ini file.  Of course, the production OAuth client would eventually be created, but I don't think we need to create an official Code For Greenville Events API client until we're closer to deploying changes to the live server.

I'm not entirely sure how expiration values and refreshing tokens works from a server to server flow, but I imagine it's not difficult.

Once the auth token stufff is taken care of, we likely need to do a [query to groupByUrlname](https://www.meetup.com/api/schema/#groupByUrlname) which returns a [Group](https://www.meetup.com/api/schema/#Group) object, which has a upcomingEvents response as an [UpcomingEventsConnection object](https://www.meetup.com/api/schema/#UpcomingEventsConnection), which then has pagination and edge details to help iterate through the list of [UpcomingEventsEdge objects](https://www.meetup.com/api/schema/#UpcomingEventsEdge), which contains nodes as an [Event object](https://www.meetup.com/api/schema/#Event).

**Comment by allella** [comment]
Not sure if this will be needed, but I did see that the Wayback Machine has copies of the old v2 REST API docs, which were removed and redirected to the GraphQL docs awhile back.

[v2 REST API - GET /events](https://web.archive.org/web/20170709041824/http://www.meetup.com/meetup_api/docs/2/events/)
[v2 REST API - GET /groups](https://web.archive.org/web/20170709041556/http://www.meetup.com/meetup_api/docs/2/groups/)

And, yes, I have donated to Wayback Machine recently, because it does things like this for you.

**Comment by allella** [comment]
[RefactorGVL now has a Meetup Pro account](https://www.meetup.com/pro/refactorgvl/). 

This means the Meetup GraphQL API migration mentioned in the first post on this issue is likely possible.

We'd have to satisfy their [Terms of Service](https://help.meetup.com/hc/en-us/articles/360027447252-Terms-of-Service) and [API License](https://help.meetup.com/hc/en-us/articles/360028705532-Meetup-API-license-terms), but hopefully that's not a blocker since we're not "commercial".  It does expect that we'd have a privacy policy defined that

> If making an application available, provide and comply with a user agreement and a privacy policy that describes how you collect, share, use, and retain user information. Your privacy policy must be at least as protective of the personal information of Meetup’s members as is Meetup’s [Privacy Policy](https://www.meetup.com/privacy/);

[A privacy policy is on the meta list.](https://github.com/hackgvl/meta/issues/25)

This may be an [older doc, but it has some FAQs on the GraphQL version of their API](https://help.meetup.com/hc/en-us/articles/360028901812-Using-Meetup-s-API).

**Comment by allella** [comment]
@irby thanks for jumping on this one.

The notes above should be a good start.  @raiderrobert controls the Meetup Pro account, which it seems is necessary to create an eventual OAuth needed to hit the Meetup GraphQL API.

**Comment by allella** [comment]
@raiderrobert @irby even with Meetup pricing increases mentioned in the community-organizers channel, it doesn't seem we're going to replace Meetup soon enough to continue hoping the REST API works.

The Refactor _Pro_ account will likely be needed if we're going to try to authenticate to Meetup's GraphQL API.

We're not trying to create a client, since we're going server to server.  So, I think [the JWT Flow in the docs is the way we'd have to authenticate](https://www.meetup.com/api/authentication/#p04-jwt-flow-section).

Robert, it looks like either of the following links may have an interface to CRUD clients and gather more context for considering this issue
https://www.meetup.com/api/oauth/list/
https://www.meetup.com/api/oauth/create/

**Comment by irby** [comment]
@allella I spoke with @raiderrobert yesterday. Yes, a Pro account is required in order to create an OAuth client, and we will need a client in order to generate the JWT necessary for JWT flow.

Robert submitted an application to create an OAuth client yesterday. It took a couple days for my application to be reviewed (and rejected because I wasn't Pro), so I expect it won't be for another couple days that we'll get a client together.

I have already made great strides on developing the code with what I've been able to get from the GraphQL playground. @bogdankharchenko was a tremendous help in getting me started.
